File: apps/api/src/presentation/api/v1/student.py
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, BackgroundTasks, Depends, HTTPException, Query, Request, status

from src.application.dtos.learning import (
    LearningAnalyticsResponse,
    MasteryRecordResponse,
    RecommendationResponse,
    SessionListResponse,
    StudentProfileResponse,
    UpdateProfileRequest,
)
from src.application.dtos.user import UserResponse
from src.application.interfaces.dependencies import (
    get_current_user,
    get_learning_analytics_service,
    get_learning_gap_detector,
    get_mastery_engine,
    get_next_best_topic_engine,
    get_session_memory_service,
    get_student_profile_service,
)
from src.application.services.learning_analytics_service import LearningAnalyticsService
from src.application.services.learning_gap_detector import LearningGapDetector
from src.application.services.mastery_engine import MasteryEngine
from src.application.services.next_best_topic_engine import NextBestTopicEngine
from src.application.services.session_memory_service import SessionMemoryService
from src.application.services.student_profile_service import StudentProfileService
from src.presentation.api.response import ok, paginated

router = APIRouter(prefix="/student", tags=["Learning Engine"])


@router.get("/profile", response_model=None)
async def get_profile(
    request: Request,
    current_user: Annotated[UserResponse, Depends(get_current_user)],
    profile_svc: Annotated[StudentProfileService, Depends(get_student_profile_service)],
):
    profile = await profile_svc.get_or_create(UUID(current_user.id))
    return ok(
        StudentProfileResponse(
            id=profile.id,
            user_id=profile.user_id,
            institution=profile.institution,
            grade=profile.grade,
            subjects=profile.subjects,
            current_chapter=profile.current_chapter,
            learning_velocity=profile.learning_velocity,
            confidence_score=profile.confidence_score,
            last_active=profile.last_active,
            created_at=profile.created_at,
            updated_at=profile.updated_at,
        ).model_dump(),
        request_id=request.state.request_id,
    )


@router.post("/profile", response_model=None)
async def update_profile(
    body: UpdateProfileRequest,
    request: Request,
    current_user: Annotated[UserResponse, Depends(get_current_user)],
    profile_svc: Annotated[StudentProfileService, Depends(get_student_profile_service)],
):
    profile = await profile_svc.update(
        user_id=UUID(current_user.id),
        institution=body.institution,
        grade=body.grade,
        subjects=body.subjects or None,
        current_chapter=body.current_chapter,
    )
    return ok(
        StudentProfileResponse(
            id=profile.id,
            user_id=profile.user_id,
            institution=profile.institution,
            grade=profile.grade,
            subjects=profile.subjects,
            current_chapter=profile.current_chapter,
            learning_velocity=profile.learning_velocity,
            confidence_score=profile.confidence_score,
            last_active=profile.last_active,
            created_at=profile.created_at,
            updated_at=profile.updated_at,
        ).model_dump(),
        request_id=request.state.request_id,
    )


@router.get("/sessions", response_model=None)
async def list_sessions(
    request: Request,
    current_user: Annotated[UserResponse, Depends(get_current_user)],
    session_svc: Annotated[SessionMemoryService, Depends(get_session_memory_service)],
    page: int = Query(default=1, ge=1),
    limit: int = Query(default=20, ge=1, le=100),
):
    offset = (page - 1) * limit
    sessions, total = await session_svc.list_sessions(
        UUID(current_user.id), limit=limit, offset=offset
    )
    return paginated(
        data=[
            {
                "id": str(s.id),
                "user_id": str(s.user_id),
                "conversation_id": str(s.conversation_id) if s.conversation_id else None,
                "subject": s.subject,
                "chapter": s.chapter,
                "grade": s.grade,
                "question": s.question,
                "primary_concept": s.primary_concept,
                "concepts_discussed": s.concepts_discussed,
                "bloom_level": s.bloom_level,
                "difficulty_level": s.difficulty_level,
                "misconceptions": s.misconceptions,
                "token_count": s.token_count,
                "duration_ms": s.duration_ms,
                "created_at": s.created_at.isoformat(),
            }
            for s in sessions
        ],
        total=total,
        page=page,
        limit=limit,
        request_id=request.state.request_id,
    )


@router.get("/mastery", response_model=None)
async def get_mastery(
    request: Request,
    current_user: Annotated[UserResponse, Depends(get_current_user)],
    mastery_engine: Annotated[MasteryEngine, Depends(get_mastery_engine)],
):
    records = await mastery_engine.get_all(UUID(current_user.id))
    return ok(
        [
            MasteryRecordResponse(
                id=r.id,
                concept_id=r.concept_id,
                concept_name=r.concept_name,
                score=r.score,
                label=r.label,
                interaction_count=r.interaction_count,
                last_updated=r.last_updated,
            ).model_dump()
            for r in records
        ],
        request_id=request.state.request_id,
    )


@router.get("/gaps", response_model=None)
async def get_gaps(
    request: Request,
    current_user: Annotated[UserResponse, Depends(get_current_user)],
    gap_detector: Annotated[LearningGapDetector, Depends(get_learning_gap_detector)],
    include_resolved: bool = Query(default=False),
):
    gaps = await gap_detector.list_gaps(
        UUID(current_user.id), include_resolved=include_resolved
    )
    return ok(
        [
            {
                "id": str(g.id),
                "concept_id": str(g.concept_id),
                "concept_name": g.concept_name,
                "severity": g.severity,
                "reason": g.reason,
                "confidence": g.confidence,
                "occurrence_count": g.occurrence_count,
                "is_resolved": g.is_resolved,
                "created_at": g.created_at.isoformat(),
                "updated_at": g.updated_at.isoformat(),
            }
            for g in gaps
        ],
        request_id=request.state.request_id,
    )


@router.post("/gaps/{gap_id}/resolve", response_model=None)
async def resolve_gap(
    gap_id: UUID,
    request: Request,
    current_user: Annotated[UserResponse, Depends(get_current_user)],
    gap_detector: Annotated[LearningGapDetector, Depends(get_learning_gap_detector)],
):
    await gap_detector.resolve(gap_id)
    return ok({"resolved": True}, request_id=request.state.request_id)


@router.get("/recommendations", response_model=None)
async def get_recommendations(
    request: Request,
    current_user: Annotated[UserResponse, Depends(get_current_user)],
    engine: Annotated[NextBestTopicEngine, Depends(get_next_best_topic_engine)],
):
    recs = await engine.recommend(UUID(current_user.id))
    return ok(
        [
            RecommendationResponse(
                concept_id=r.concept_id,
                concept_name=r.concept_name,
                subject=r.subject,
                chapter=r.chapter,
                reason=r.reason,
                readiness_score=r.readiness_score,
            ).model_dump()
            for r in recs
        ],
        request_id=request.state.request_id,
    )


@router.get("/analytics", response_model=None)
async def get_analytics(
    request: Request,
    current_user: Annotated[UserResponse, Depends(get_current_user)],
    analytics_svc: Annotated[LearningAnalyticsService, Depends(get_learning_analytics_service)],
):
    analytics = await analytics_svc.get_analytics(UUID(current_user.id))
    return ok(
        {
            "user_id": str(analytics.user_id),
            "total_sessions": analytics.total_sessions,
            "total_concepts_encountered": analytics.total_concepts_encountered,
            "average_mastery": analytics.average_mastery,
            "mastered_count": analytics.mastered_count,
            "developing_count": analytics.developing_count,
            "emerging_count": analytics.emerging_count,
            "not_started_count": analytics.not_started_count,
            "active_gaps": analytics.active_gaps,
            "critical_gaps": analytics.critical_gaps,
            "recent_bloom_levels": analytics.recent_bloom_levels,
            "recent_concepts": analytics.recent_concepts,
        },
        request_id=request.state.request_id,
    )

File: apps/web/lib/api/student.ts
import { apiClient } from './client'

export interface BehavioralSignals {
  preferred_bloom_level: string | null
  struggle_bloom_level: string | null
  avg_session_duration_ms: number
  sessions_per_day: number
  question_complexity_trend: 'rising' | 'stable' | 'declining'
  dominant_subject: string | null
  total_sessions: number
  total_misconceptions: number
  engagement_streak: number
  strengths: string[]
  recent_topics: string[]
  response_pattern: 'procedural' | 'conceptual' | 'mixed' | 'unknown'
}

export interface StudentProfile {
  id: string
  user_id: string
  institution: string | null
  grade: string | null
  subjects: string[]
  current_chapter: string | null
  learning_velocity: number
  confidence_score: number
  behavioral_signals: BehavioralSignals
  last_active: string
  created_at: string
  updated_at: string
}

export interface UpdateProfilePayload {
  institution?: string | null
  grade?: string | null
  subjects?: string[]
  current_chapter?: string | null
}

// ── Psychometric (Phase A) ────────────────────────────────────────────────

export interface PsychometricQuestionOption {
  value: string
  label: string
}

export interface PsychometricQuestion {
  key: string
  dimension: 'motivation' | 'discipline' | 'interest' | 'learning_style' | 'confidence'
  type: 'likert' | 'choice'
  text: string
  options: PsychometricQuestionOption[]
}

export interface PsychometricProfile {
  motivation_type: 'intrinsic' | 'extrinsic' | 'mixed' | null
  motivation_strength: number
  discipline: number
  interests: string[]
  learning_style_preference: 'visual' | 'verbal' | 'hands_on' | 'reading' | null
  confidence_self_report: number
  sources: Record<string, 'survey' | 'inferred' | 'blended'>
  completeness: number
  assessed_at: string | null
}

export interface LearningSession {
  id: string
  user_id: string
  conversation_id: string | null
  subject: string | null
  chapter: string | null
  grade: string | null
  question: string
  primary_concept: string | null
  concepts_discussed: string[]
  bloom_level: string
  difficulty_level: string
  misconceptions: string[]
  intent: 'concept_explanation' | 'problem_solving' | 'clarification' | 'recall' | 'test_prep' | 'correction_request' | 'unknown'
  token_count: number
  duration_ms: number
  created_at: string
}

export interface ConceptMemory {
  id: string
  concept_id: string
  concept_name: string
  times_taught: number
  successful_approaches: number
  failed_approaches: number
  last_approach: string | null
  teaching_notes: string[]
  success_rate: number
  needs_different_approach: boolean
  last_taught: string
}

export interface LearningPathNode {
  concept_id: string
  concept_name: string
  subject: string | null
  chapter: string | null
  bloom_level: string
  difficulty: string
}

export interface LearningPath {
  target_concept_id: string | null
  path: LearningPathNode[]
  frontier: Recommendation[]
  coverage: Record<string, { total: number; mastered: number; coverage_pct: number }>
}

export interface SkillEntry {
  skill: string
  proficiency: number
}

export interface SkillProfile {
  skill_profile: Record<string, number>
  top_skills: SkillEntry[]
  bloom_distribution: Record<string, number>
}

export interface MasteryRecord {
  id: string
  concept_id: string
  concept_name: string
  score: number
  label: 'mastered' | 'developing' | 'emerging' | 'not_started'
  interaction_count: number
  last_updated: string
}

export interface LearningGap {
  id: string
  concept_id: string
  concept_name: string
  severity: 'low' | 'medium' | 'high' | 'critical'
  reason: string
  confidence: 'low' | 'medium' | 'high'
  occurrence_count: number
  is_resolved: boolean
  created_at: string
  updated_at: string
}

export interface Recommendation {
  concept_id: string
  concept_name: string
  subject: string | null
  chapter: string | null
  reason: string
  readiness_score: number
}

export interface LearningAnalytics {
  user_id: string
  total_sessions: number
  total_concepts_encountered: number
  average_mastery: number
  mastered_count: number
  developing_count: number
  emerging_count: number
  not_started_count: number
  active_gaps: number
  critical_gaps: number
  recent_bloom_levels: Record<string, number>
  recent_concepts: string[]
  velocity_trend: number
  at_risk_concepts: RetentionRisk[]
}

export interface RetentionRisk {
  concept_id: string
  concept_name: string
  score: number
  days_since_reinforced: number
  risk: number
}

export interface SessionsPage {
  data: LearningSession[]
  total: number
  page: number
  limit: number
}

export const studentApi = {
  getProfile: () => apiClient.get<StudentProfile>('/api/v1/student/profile'),

  updateProfile: (payload: UpdateProfilePayload) =>
    apiClient.post<StudentProfile>('/api/v1/student/profile', payload),

  getSessions: (page = 1, limit = 20) =>
    apiClient.get<SessionsPage>(`/api/v1/student/sessions?page=${page}&limit=${limit}`),

  getMastery: () => apiClient.get<MasteryRecord[]>('/api/v1/student/mastery'),

  getGaps: (includeResolved = false) =>
    apiClient.get<LearningGap[]>(
      `/api/v1/student/gaps?include_resolved=${includeResolved}`
    ),

  resolveGap: (gapId: string) =>
    apiClient.post<{ resolved: boolean }>(`/api/v1/student/gaps/${gapId}/resolve`, {}),

  getRecommendations: () =>
    apiClient.get<Recommendation[]>('/api/v1/student/recommendations'),

  getAnalytics: () => apiClient.get<LearningAnalytics>('/api/v1/student/analytics'),

  getConceptMemory: () => apiClient.get<ConceptMemory[]>('/api/v1/student/memory'),

  getLearningPath: (targetConceptId?: string) =>
    apiClient.get<LearningPath>(
      `/api/v1/student/learning-path${targetConceptId ? `?target_concept_id=${targetConceptId}` : ''}`
    ),

  getSkillProfile: () => apiClient.get<SkillProfile>('/api/v1/student/skills'),

  // ── Psychometric (Phase A) ────────────────────────────────────────────────

  getPsychometricQuestions: (limit = 5) =>
    apiClient.get<PsychometricQuestion[]>(
      `/api/v1/student/psychometric/questions?limit=${limit}`
    ),

  submitPsychometricResponse: (payload: { question_key: string; value: string }) =>
    apiClient.post<PsychometricProfile>('/api/v1/student/psychometric/responses', payload),

  getPsychometricProfile: () =>
    apiClient.get<PsychometricProfile>('/api/v1/student/psychometric/profile'),

  // ── Quiz & Assessment ───────────────────────────────────────────────────

  generateQuiz: (body: {
    subject?: string | null
    chapter?: string | null
    concept_ids?: string[] | null
    question_count?: number
    difficulty?: string
  }) => apiClient.post<import('@roognis/shared').QuizSummaryDto>('/api/v1/student/quiz/generate', body),

  // Backend paginated() puts the array at top-level `data` (pagination in
  // `meta`), so apiClient.get returns ApiResponse<QuizSummaryDto[]>.
  getQuizzes: (page = 1, limit = 20, subject?: string) =>
    apiClient.get<import('@roognis/shared').QuizSummaryDto[]>(
      `/api/v1/student/quiz?page=${page}&limit=${limit}${subject ? `&subject=${encodeURIComponent(subject)}` : ''}`
    ),

  getQuiz: (quizId: string) =>
    apiClient.get<import('@roognis/shared').QuizDetailDto>(`/api/v1/student/quiz/${quizId}`),

  submitQuiz: (quizId: string, responses: import('@roognis/shared').SubmitResponseDto[]) =>
    apiClient.post<import('@roognis/shared').QuizAttemptDto>(`/api/v1/student/quiz/${quizId}/submit`, { responses }),

  getAttemptResults: (attemptId: string) =>
    apiClient.get<import('@roognis/shared').QuizResultDto>(`/api/v1/student/quiz/attempts/${attemptId}/results`),

  getQuizHistory: (limit = 10) =>
    apiClient.get<import('@roognis/shared').QuizHistoryItemDto[]>(`/api/v1/student/quiz/history?limit=${limit}`),
}
